framework/src/org/checkerframework/common/value/ValueTransfer.java

```diff
@@ -521,90 +597,165 @@ private AnnotationMirror createNumberAnnotationMirror(List<Number> values) {
             Node leftNode,
             Node rightNode,
             ComparisonOperators op,
-            TransferInput<CFValue, CFStore> p) {
+            TransferInput<CFValue, CFStore> p,
+            CFStore thenStore,
+            CFStore elseStore) {
         List<? extends Number> lefts = getNumericalValues(leftNode, p);
         List<? extends Number> rights = getNumericalValues(rightNode, p);
         if (lefts == null || rights == null) {
             return null;
         }
         List<Boolean> resultValues = new ArrayList<>();
+
+        // These lists are used to refine the values in the store based on the results of the comparison.
+        List<Number> thenLeftVals = new ArrayList<>();
+        List<Number> elseLeftVals = new ArrayList<>();
+        List<Number> thenRightVals = new ArrayList<>();
+        List<Number> elseRightVals = new ArrayList<>();
+
         for (Number left : lefts) {
             NumberMath<?> nmLeft = NumberMath.getNumberMath(left);
             for (Number right : rights) {
+                Boolean result;
                 switch (op) {
                     case EQUAL:
-                        resultValues.add(nmLeft.equalTo(right));
+                        result = nmLeft.equalTo(right);
                         break;
                     case GREATER_THAN:
-                        resultValues.add(nmLeft.greaterThan(right));
+                        result = nmLeft.greaterThan(right);
                         break;
                     case GREATER_THAN_EQ:
-                        resultValues.add(nmLeft.greaterThanEq(right));
+                        result = nmLeft.greaterThanEq(right);
                         break;
                     case LESS_THAN:
-                        resultValues.add(nmLeft.lessThan(right));
+                        result = nmLeft.lessThan(right);
                         break;
                     case LESS_THAN_EQ:
-                        resultValues.add(nmLeft.lessThanEq(right));
+                        result = nmLeft.lessThanEq(right);
                         break;
                     case NOT_EQUAL:
-                        resultValues.add(nmLeft.notEqualTo(right));
+                        result = nmLeft.notEqualTo(right);
                         break;
                     default:
                         throw new UnsupportedOperationException();
                 }
+                resultValues.add(result);
+                if (result) {
+                    thenLeftVals.add(left);
+                    thenRightVals.add(right);
+                } else {
+                    elseLeftVals.add(left);
+                    elseRightVals.add(right);
+                }
             }
         }
+
+        // The result can only be bottom if both the comparison had some results and one of the four lists is empty.
+        boolean canGoToBottom = resultValues.size() > 0;
```

Comment:
I've taken a look at the test case you provided. After quite a bit of digging, it turns out that the problem is that the Value Checker isn't handling the mix of `Integer` and `int` quite as precisely as we would like. Here are two simplified test cases:

```java
    void simpleTest1(@BottomVal Integer x, int y) {
	if (x == y) {
	    @BottomVal int z = y;
	}
    }

    void simpleTest2(@BottomVal Integer x, Integer y) {
	if (x == y) {
	    @BottomVal int z = y;
	}
    }
```

`SimpleTest1` issues a compile warning on the assignment to `@BottomVal int z`. `SimpleTest2` does not. I've determined that this happens because when an `Integer` and an `int` are compared, the `Integer`, `x`, is desugared to `x.intValue()`, which is compared to `y`. But `x.intValue()`'s type isn't being properly assigned when `x` is `@BottomVal`. 

I can look for a fix, but I don't think it'll be in the code here, and I wanted to let you know what was going on.